### PR TITLE
Enable error-recovery and reporting support

### PR DESCRIPTION
**Pull request by hershi**
[ see #675 ]
Added support for the caller to pass in an error handler that would receive reports of errors, and can indicate whether we want to recover from the errors or stop evaluation.
With this support in place, we can now recover from some errors, thus reporting multiple errors per run in stead of just the first one, which can be pretty tedious.

This change include support for recovering from one specific error - instantiating an array with an abstract value for length. Enabling recovery from other errors will be done separately (and gradually).

Comments from Herman's initial review were addressed (thanks @hermanventer!)

### CODE REVIEW

File: src/errors.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    this.stack = stack;
+  }
+}
+
+// This error is used to indicate a failure due to previously encountered
+// errors that were deferred by the error-handler. The original errors
+// have already been reported to the error-handler so their details are not
+// included in this error (it's the responsibility of the error-handler to
+// track them if needed)
+export class DeferredErrorsError {
```
Comment (hermanventer) [inline src/errors.js]: Why does this not extend Error?

File: src/intrinsics/ecma262/Array.js [hunk 2]
```diff
@@ -85,7 +86,12 @@ export default function (realm: Realm): NativeFunctionValue {
         // c. Let intLen be 1.
         intLen = 1;
       } else { // 7. Else,
-        len = len.throwIfNotConcreteNumber();
+        // len = recoverableError(() => len.throwIfNotConcreteNumber(), ()=>new NumberValue(realm, 0));
+        len = recoverableError(
+          realm,
+          ()=>len.throwIfNotConcreteNumber(),
+          ()=>new NumberValue(realm, 0));
+
```
Comment (hermanventer) [inline src/intrinsics/ecma262/Array.js]: Hmm. First you write the entire statement in one line, but in the form of a comment. Then you repeat it verbatim, but you break it up into multiple lines.

I'm also not thrilled with the helper function. I'd rather see something like "if (something is bad && reportError("Hey, this thing is bad because it's BAD") === "recover") { do some recovery thing };".

File: src/errors.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/* @flow */
+
+// This is the error format used to report errors to the caller-supplied
+// error-handler
+export class NativeError extends Error {
```
Comment (hermanventer) [inline src/errors.js]: What makes this a "native" error?

File: src/prepack-cli.js [hunk 4]
```diff
@@ -129,6 +135,12 @@ function run(Object, Array, console, JSON, process, prepackStdin, prepackFileSyn
         // their errors to the console, but exit with an error code.
         process.exit(1);
       }
+      if (x instanceof DeferredErrorsError) {
+        // Report the general failure reason. The specific errors were already
+        // handled and reported by the error-handler
+        console.log(x.message);
```
Comment (hermanventer) [inline src/prepack-cli.js]: It is somewhat unusual for a compiler to report "No object code was generated because there were compilation errors" and this seems sufficiently analogous that I wonder why this is necessary.

File: src/realm.js [hunk 5]
```diff
@@ -614,4 +621,25 @@ export class Realm {
     }
     for (; i < generatorBody.length; i++) realmGeneratorBody.push(generatorBody[i]);
   }
+
+  handleError(error: IntrospectionThrowCompletion): boolean {
```
Comment (hermanventer) [inline src/realm.js]: We very likely will want to also handle errors that are not Introspection errors.

File: src/realm.js [hunk 6]
```diff
@@ -614,4 +621,25 @@ export class Realm {
...
+    // It's important to accumulate the erors so that even if there's a handler
+    // and we recover-and-continue, we can still fail the run at the end and
+    // report the errors we found.
```
Comment (hermanventer) [inline src/realm.js]: The comment seems a bit outdated and not all that helpful.

File: src/realm.js (hunk 5 above)
Comment (hermanventer) [inline src/realm.js]: How about a comment that explains the meaning of the result?

File: src/serializer/serializer.js [hunk 7]
```diff
@@ -171,6 +171,7 @@ export class Serializer {
         if (onError) {
           onError(realm, res.value);
         }
+        console.log("Execution returned a completion!");
```
Comment (hermanventer) [inline src/serializer/serializer.js]: We are already logging something. Logging it (possibly again) to the console seems uncalled for.

File: src/errors.js (hunk 1 above)
Comment (hershi) [inline src/errors.js]: By mistake. But I removed it altogether now based on your other comment

File: src/prepack-cli.js (hunk 4 above)
Comment (hershi) [inline src/prepack-cli.js]: Removed

File: src/realm.js (hunk 5 above)
Comment (hershi) [inline src/realm.js]: I thought we wanted to limit it to that for now, no? It's true that we may want to handle other errors in the future, but would be hard to write generic code to convert them to native. We can adjust this handler as needed in the future, since it's not used by the external client (that is abstracted in the NativeError).

File: src/realm.js [hunk 8]
```diff
@@ -619,4 +626,25 @@ export class Realm {
     }
     for (; i < generatorBody.length; i++) realmGeneratorBody.push(generatorBody[i]);
   }
+
+  handleError(error: IntrospectionThrowCompletion): boolean {
+    // It's important to accumulate the erors so that even if there's a handler
```
Comment (NTillmann) [inline src/realm.js]: spelling

File: src/intrinsics/ecma262/Array.js (hunk 2 above)
Comment (hershi) [inline src/intrinsics/ecma262/Array.js]: The comment was a remnant from when I was trying to model the helper function.
Regarding the helper function, my goal was to avoid requiring overly verbose code, since the way the code is currently structured relies on exceptions - which means that ```something is bad``` from above would translate into a try-catch
My original version had that try-catch explicitly in the code, and I introduced the helper function to make it easy to add support for more recoverable errors. I'm open to suggestions regarding other ways to handle this, or if you think writing the verbose version of the code wherever we want to enable handling recovery would be more fitting.

File: src/intrinsics/ecma262/Error.js [hunk 9]
```diff
@@ -74,10 +74,10 @@ function buildStack(realm: Realm, context: ObjectValue) {
     msg.throwIfNotConcrete();
   }
 
-  for (let executionContext of stack) {
+  for (let executionContext of stack.reverse()) {
     let caller = executionContext.caller;
     let locString = describeLocation(realm, caller ? caller.function : undefined, caller ? caller.lexicalEnvironment : undefined, executionContext.loc);
-    if (locString !== undefined) lines.unshift(locString);
+    if (locString !== undefined) lines.push(locString);
```
Comment (NTillmann) [inline src/intrinsics/ecma262/Error.js]: Is reverse + push known to be faster, or why this change?

File: src/options.js [hunk 10]
```diff
@@ -31,6 +32,7 @@ export type Options = {|
   residual?: boolean,
   serialize?: boolean,
   strictlyMonotonicDateNow?: boolean,
+  errorHandler?: ErrorHandler,
```
Comment (NTillmann) [inline src/options.js]: Design note: `Options` used to be a pure datastructure, with the intention that it's configured on a command-line, with no behavior attached to it. I can see how it is convenient to use it for this handler, but it's a bit weird.

File: src/intrinsics/ecma262/Array.js [hunk 11]
```diff
@@ -85,7 +86,12 @@ export default function (realm: Realm): NativeFunctionValue {
         // c. Let intLen be 1.
         intLen = 1;
       } else { // 7. Else,
-        len = len.throwIfNotConcreteNumber();
+        // len = recoverableError(() => len.throwIfNotConcreteNumber(), ()=>new NumberValue(realm, 0));
+        len = recoverableError(
```
Comment (NTillmann) [inline src/intrinsics/ecma262/Array.js]: is this just to get a pull request out?
What's the bigger vision of throwing exceptions vs recoverable errors?
How is it going to be tested?

File: src/serializer/serializer.js (hunk 7 above)
Comment (NTillmann) [inline src/serializer/serializer.js]: And quite confusing to the end-user.

File: src/prepack-cli.js [hunk 12]
```diff
@@ -111,12 +112,16 @@ function run(Object, Array, console, JSON, process, prepackStdin, prepackFileSyn
     },
     flags
   );
+  resolvedOptions.errorHandler =
+    (error: NativeError) => {
+      console.log(`${error.message}\n${error.stack}`);
```
Comment (NTillmann) [inline src/prepack-cli.js]: So the individual errors are logged here. Is this going to be a good experience? Will the end-user understand what's happening if the errors stem from some particular partial evaluation branch?

File: src/errors.js (hunk 3 above)
Comment (hershi) [inline src/errors.js]: The fact that its data lives in the context of prepack/the caller rather than the context of the program being run.
I've changed it to ProgramEvaluationError, though I'm not sure if that's much better... let me know if you have other suggestions

File: src/intrinsics/ecma262/Array.js (hunk 11 above)
Comment (hershi) [inline src/intrinsics/ecma262/Array.js]: Based on discussion with Herman, the plan is to focus this change on introducing the mechanism needed for recovering from errors, while leveraging it in specific places where we want to support recovery will be done separately (and probably gradually).

Using the new mechanism in this location is intended to:
1. Serve as a sanity check + usage example
2. Allow for testing (I'm working on the tests and they'll be included in the next iteration of the review)

File: src/intrinsics/ecma262/Error.js (hunk 9 above)
Comment (hershi) [inline src/intrinsics/ecma262/Error.js]: "Known" might be a strong word, since I'm very new to JS :)
I'm assuming that arrays behave more or less like Arrays/Vectors in other languages. If that's the case, then calling ```unshift``` n times will result in O(n^2), because for every call we are copying all the previously inserted items, where's pushing would be O(n), with reverse taking either O(n) (if it actually creates a new copy) or O(1) (if it works like an iterator adapter, though my guess was that in JS it's the former anyway), with a total complexity of O(n).

Are my assumptions regarding JS wrong here or anything else I'm missing?

Comment (hershi) [inline src/intrinsics/ecma262/Error.js]: I've also run a minor benchmark and my assumptions seem to hold - considerably faster to push vs. unshift

File: src/options.js (hunk 10 above)
Comment (hershi) [inline src/options.js]: Ah, I didn't know that bit. I'll look at restructuring it.

File: src/realm.js (hunk 5 above)
Comment (hershi) [inline src/realm.js]: Done

File: src/realm.js (hunk 8 above)
Comment (hershi) [inline src/realm.js]: Done

File: src/realm.js (hunk 6 above)
Comment (hershi) [inline src/realm.js]: Removed

File: src/serializer/serializer.js (hunk 7 above)
Comment (hershi) [inline src/serializer/serializer.js]: Removed

File: src/prepack-cli.js (hunk 12 above)
Comment (hershi) [inline src/prepack-cli.js]: Will stop by to understand the concern better

File: src/environment.js [hunk 13]
```diff
@@ -13,6 +13,7 @@ import type { BabelNode, BabelNodeFile } from "babel-types";
 import type { NormalCompletion } from "./completions.js";
 import type { Realm } from "./realm.js";
 import type { SourceType } from "./types.js";
+import { DeferredErrorsError } from "./errors.js";
```
Comment (hershi) [inline src/environment.js]: Removed

**Comment by hershi** [comment]
After further discussion with Herman, we decided to:
1. Focus this change on the public-facing API and hone the type definitions there
2. While we introduce the notion of recovery in the API (specifically, in the handler return value), not implement recovery anywhere in the code yet
3. Implement calling to the handler in a single location in the code as a sanity check for wiring the API (but, further to no. 2 above, fail in that location, regardless of handler result)

Next iteration will reflect the above

File: src/options.js (hunk 10 above)
Comment (hershi) [inline src/options.js]: I've changed this to be passed independently of Options in prepack-standalone, which is needed for tests.
To make it useful for prepack-cli or any other consumer, this will need to be added to prepack-node.js as well. It's easy to do, but since it's not very useful at the moment, I figured it's better to leave it out for now to not mess up the interface with unnecessary parameters for now. Let me know if you think otherwise and want me to introduce it there as well.

File: src/errors.js [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+/* @flow */
+
+export type Severity = 'Error' | 'Warning';
+export type ErrorHandlerResult = 'Fail' | 'RecoverIfPossible';
+export type ErrorCode = 'PP0001';
+
+// This is the error format used to report errors to the caller-supplied
+// error-handler
+export class CompilerDiagnostics extends Error {
+  constructor(message: string, location: string, severity: Severity = 'Error', errorCode: string) {
```
Comment (hermanventer) [inline src/errors.js]: Parameter with default values should appear last in the list.

File: test/error-handler/testErrorHandlerCalled.js [hunk 15]
```diff
@@ -0,0 +1,15 @@
+// expected errors: 1
+
+var a;
+//var b = __abstract("boolean");
```
Comment (hermanventer) [inline test/error-handler/testErrorHandlerCalled.js]: Don't leave commented out code in the source.

File: test/error-handler/testErrorHandlerCalled.js [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+var b = true;
+if(b) {
+  var l = __abstract("number");
+  a = new Array(l);
+  a = new Array(l);
+  a = new Array(l);
+}
+else {
```
Comment (hermanventer) [inline test/error-handler/testErrorHandlerCalled.js]: This does not follow our formatting conventions.

File: test/error-handler/testErrorHandlerCalled.js [hunk 17]
```diff
@@ -0,0 +1,15 @@
+// expected errors: 1
```
Comment (hermanventer) [inline test/error-handler/testErrorHandlerCalled.js]: It would be better to have a way of checking the content of the error.

File: src/errors.js [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+    super(message);
+
+    this.location = location;
+    this.severity = severity;
+    this.errorCode = errorCode;
+  }
+
+  location: string;
```
Comment (hermanventer) [inline src/errors.js]: Rather make this be of type BabelNodeSourceLocation.

File: src/errors.js (hunk 14 above)
Comment (hershi) [inline src/errors.js]: Done

File: src/errors.js (hunk 18 above)
Comment (hershi) [inline src/errors.js]: Done

File: test/error-handler/testErrorHandlerCalled.js (hunk 17 above)
Comment (hershi) [inline test/error-handler/testErrorHandlerCalled.js]: Done

File: test/error-handler/testErrorHandlerCalled.js (hunk 15 above)
Comment (hershi) [inline test/error-handler/testErrorHandlerCalled.js]: Done

File: test/error-handler/testErrorHandlerCalled.js (hunk 16 above)
Comment (hershi) [inline test/error-handler/testErrorHandlerCalled.js]: Done

File: package.json [hunk 19]
```diff
@@ -35,6 +35,7 @@
     "test-test262": "babel-node scripts/test262-runner.js",
     "test-test262-nightly": "NIGHTLY_BUILD=true babel-node -- scripts/test262-runner.js",
     "test-internal": "babel-node --stack_size=10000 --stack_trace_limit=200 --max_old_space_size=8192 scripts/test-internal.js",
+    "test-error-handler": "babel-node --stack_size=10000 --stack_trace_limit=200 --max_old_space_size=8192 scripts/test-error-handler.js",
```
Comment (hermanventer) [inline package.json]: Please add "npm run test-error-handler" to the "test" entry below.

Comment (hermanventer) [inline package.json]: Also add a corresponding entry to circle.yml.

**Comment by hermanventer** [comment]
Getting close, but please attend to the test runners.

**Comment by hershi** [comment]
Added the test runners and validated that it runs in both the local run (npm run test) and in the circleCI report.
